File: clients/ios/AwareChat-iOS/AwareChat-iOS/Core/Networking/WebSocketClient.swift
import Foundation

nonisolated enum WebSocketConnectionState: Equatable, Sendable {
  case disconnected
  case connecting
  case connected
  case connectionFailure
}

protocol WebSocketClientProtocol: Sendable {
  func connect() async -> AsyncThrowingStream<ServerEvent, any Error>
  func send(_ event: ClientEvent) async throws
  func disconnect() async
  func connectionState() async -> WebSocketConnectionState
}

actor WebSocketClient: WebSocketClientProtocol {
  private let configuration: NetworkConfiguration
  private let factory: any WebSocketTransportFactory
  private let codec: ProtocolCodec

  private var state = WebSocketConnectionState.disconnected
  private var transport: (any WebSocketTransport)?
  private var receiveTask: Task<Void, Never>?
  private var continuation: AsyncThrowingStream<ServerEvent, any Error>.Continuation?

  init(
    configuration: NetworkConfiguration,
    factory: any WebSocketTransportFactory = URLSessionWebSocketTransportFactory(),
    codec: ProtocolCodec = ProtocolCodec()
  ) {
    self.configuration = configuration
    self.factory = factory
    self.codec = codec
  }

  func connect() -> AsyncThrowingStream<ServerEvent, any Error> {
    disconnectCurrentTransport()
    state = .connecting

    let transport = factory.makeTransport(url: configuration.webSocketURL)
    self.transport = transport

    let stream = AsyncThrowingStream<ServerEvent, any Error> { continuation in
      self.continuation = continuation
    }

    transport.start()
    state = .connected
    receiveTask = Task { [weak self] in
      await self?.receiveLoop(from: transport)
    }
    return stream
  }

  func send(_ event: ClientEvent) async throws {
    guard !Task.isCancelled else { throw NetworkError.cancelled }
    guard state == .connected, let transport else {
      throw NetworkError.notConnected
    }

    let data: Data
    do {
      data = try codec.encode(event)
    } catch {
      throw NetworkError.encoding
    }

    guard let text = String(data: data, encoding: .utf8) else {
      throw NetworkError.encoding
    }

    do {
      try await transport.send(text: text)
    } catch {
      throw NetworkError.wrapping(error)
    }
  }

  func disconnect() {
    disconnectCurrentTransport()
    state = .disconnected
  }

  func connectionState() -> WebSocketConnectionState {
    state
  }

  private func receiveLoop(from activeTransport: any WebSocketTransport) async {
    do {
      while !Task.isCancelled {
        let frame = try await activeTransport.receive()
        guard !Task.isCancelled, let transport, transport === activeTransport else { return }
        guard case .text(let text) = frame,
              let data = text.data(using: .utf8) else {
          throw NetworkError.invalidWebSocketFrame
        }
        let event: ServerEvent
        do {
          event = try codec.decodeServerEvent(from: data)
        } catch {
          throw NetworkError.decoding
        }
        continuation?.yield(event)
      }
    } catch {
      guard !Task.isCancelled, let transport, transport === activeTransport else { return }
      let networkError: NetworkError
      if let close = activeTransport.closeDetails {
        networkError = .webSocketClosed(code: close.code, reason: close.reason)
      } else {
        networkError = NetworkError.wrapping(error)
      }
      state = .connectionFailure
      continuation?.finish(throwing: networkError)
      activeTransport.cancel(code: .goingAway, reason: nil)
      clear(activeTransport: activeTransport)
    }
  }

  private func disconnectCurrentTransport() {
    receiveTask?.cancel()
    receiveTask = nil
    transport?.cancel(code: .normalClosure, reason: nil)
    transport = nil
    continuation?.finish()
    continuation = nil
  }

  private func clear(activeTransport: any WebSocketTransport) {
    guard let transport, transport === activeTransport else { return }
    self.transport = nil
    receiveTask = nil
    continuation = nil
  }
}

File: clients/android/AwareChat-Android/app/src/main/java/com/example/awarechat_android/core/network/WebSocketClient.kt
package com.example.awarechat_android.core.network

import com.example.awarechat_android.core.protocol.ClientEvent
import com.example.awarechat_android.core.protocol.ProtocolCodec
import com.example.awarechat_android.core.protocol.ServerEvent
import java.io.IOException
import java.util.UUID
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.callbackFlow

enum class WebSocketConnectionState {
    Disconnected,
    Connecting,
    Connected,
    ConnectionFailure,
}

interface WebSocketClientContract {
    val connectionState: StateFlow<WebSocketConnectionState>

    fun connect(): Flow<ServerEvent>
    suspend fun send(event: ClientEvent)
    fun disconnect()
}

class WebSocketClient(
    private val configuration: NetworkConfiguration,
    private val factory: WebSocketTransportFactory = OkHttpWebSocketTransportFactory(),
    private val codec: ProtocolCodec = ProtocolCodec(),
) : WebSocketClientContract {
    private val lock = Any()
    private val mutableConnectionState = MutableStateFlow(WebSocketConnectionState.Disconnected)
    override val connectionState: StateFlow<WebSocketConnectionState> =
        mutableConnectionState.asStateFlow()

    private var activeConnection: ActiveConnection? = null

    override fun connect(): Flow<ServerEvent> = callbackFlow {
        val id = UUID.randomUUID()
        val transport = factory.create(configuration.webSocketUrl)
        val connection = ActiveConnection(
            id = id,
            transport = transport,
            closeEvents = { error -> channel.close(error) },
            sendEvent = { event -> trySend(event) },
        )
        val replaced = synchronized(lock) {
            val previous = activeConnection
            activeConnection = connection
            mutableConnectionState.value = WebSocketConnectionState.Connecting
            previous
        }
        replaced?.terminate()

        try {
            transport.start(listener(id))
        } catch (error: Throwable) {
            fail(id, NetworkException.wrap(error))
        }

        awaitClose { stopIfActive(id) }
    }

    override suspend fun send(event: ClientEvent) {
        currentCoroutineContext().ensureActive()
        val text = try {
            codec.encode(event)
        } catch (error: CancellationException) {
            throw error
        } catch (_: Throwable) {
            throw NetworkException.Encoding()
        }

        val connection = synchronized(lock) {
            activeConnection?.takeIf {
                mutableConnectionState.value == WebSocketConnectionState.Connected
            }
        } ?: throw NetworkException.NotConnected()

        val didSend = try {
            connection.transport.send(text)
        } catch (error: CancellationException) {
            throw error
        } catch (error: IOException) {
            throw NetworkException.Transport(error)
        } catch (_: Throwable) {
            throw NetworkException.TransportFailure()
        }
        if (!didSend) throw NetworkException.NotConnected()
    }

    override fun disconnect() {
        val connection = synchronized(lock) {
            val current = activeConnection
            activeConnection = null
            mutableConnectionState.value = WebSocketConnectionState.Disconnected
            current
        }
        connection?.terminate()
    }

    private fun listener(id: UUID): WebSocketTransport.Listener =
        object : WebSocketTransport.Listener {
            override fun onOpen() {
                synchronized(lock) {
                    if (activeConnection?.id == id) {
                        mutableConnectionState.value = WebSocketConnectionState.Connected
                    }
                }
            }

            override fun onText(text: String) {
                val event = try {
                    codec.decodeServerEvent(text)
                } catch (_: Throwable) {
                    fail(id, NetworkException.Decoding())
                    return
                }
                val sendEvent = synchronized(lock) {
                    activeConnection?.takeIf { it.id == id }?.sendEvent
                }
                sendEvent?.invoke(event)
            }

            override fun onBinary(bytes: ByteArray) {
                fail(id, NetworkException.InvalidWebSocketFrame())
            }

            override fun onClosing(code: Int, reason: String) {
                fail(id, NetworkException.WebSocketClosed(code, reason.ifEmpty { null }))
            }

            override fun onClosed(code: Int, reason: String) {
                fail(id, NetworkException.WebSocketClosed(code, reason.ifEmpty { null }))
            }

            override fun onFailure(error: Throwable) {
                val failure = when (error) {
                    is IOException -> NetworkException.Transport(error)
                    else -> NetworkException.TransportFailure()
                }
                fail(id, failure)
            }
        }

    private fun fail(id: UUID, error: NetworkException) {
        val connection = synchronized(lock) {
            val current = activeConnection?.takeIf { it.id == id } ?: return
            activeConnection = null
            mutableConnectionState.value = WebSocketConnectionState.ConnectionFailure
            current
        }
        connection.transport.close(GOING_AWAY, null)
        connection.transport.cancel()
        connection.closeEvents(error)
    }

    private fun stopIfActive(id: UUID) {
        val connection = synchronized(lock) {
            val current = activeConnection?.takeIf { it.id == id } ?: return
            activeConnection = null
            mutableConnectionState.value = WebSocketConnectionState.Disconnected
            current
        }
        connection.transport.close(NORMAL_CLOSURE, null)
        connection.transport.cancel()
    }

    private data class ActiveConnection(
        val id: UUID,
        val transport: WebSocketTransport,
        val closeEvents: (Throwable?) -> Unit,
        val sendEvent: (ServerEvent) -> Unit,
    ) {
        fun terminate() {
            transport.close(NORMAL_CLOSURE, null)
            transport.cancel()
            closeEvents(null)
        }
    }

    private companion object {
        const val NORMAL_CLOSURE = 1000
        const val GOING_AWAY = 1001
    }
}
